Refactor wildcard handling

bernd opened the pull request:
- Create two PathSet types, one for globbing, one for a single path
  without globbing.
- Add "path-glob-root" and "path-glob-pattern" config options for the
  FileInput to handle the globbing case.
  The old "path" option is used when no globbing is needed.

Splitting the configuration options for globbing simplifies the implementation and makes it work on windows. Before this, we tried to detect the root path part and the pattern part of the configured file input path automatically which didn't work in all cases.

bernd committed:
Extract noop file tree walker into constant.

bernd committed:
Fix tests with jimfs usage.

bernd committed:
Refactor wildcard/globbing handling.

- Create two PathSet types, one for globbing, one for a single path
  without globbing.
- Add "path-glob-root" and "path-glob-pattern" config options for the
  FileInput to handle the globbing case.
  The old "path" option is used when no globbing is needed.

Splitting the configuration options for globbing simplifies the
implementation and makes it work on Windows. Before this, we tried to
detect the root path part and the pattern part of the configured file
input path automatically which didn't work in all cases.

joschi commented on src/main/java/org/graylog/collector/file/FileReaderService.java:
```diff
@@ -98,6 +98,11 @@ protected void doStart() {
         }
 
+        if (paths.isEmpty()) {
+            log.info("No files to follow for input \"{}\" yet, files will be followed once they appear in \"{}\"",
```
Isn't this message only true for `GlobPathSet`?

bernd replied:
No because nothing prevents you from creating an input for a file that does not exist yet.

joschi replied:
Okay, but then the message still reads strange if a single file is being used:
> No files to follow for input "Wurstsalat" yet, files will be followed once they appear in "/var/log/syslog"

joschi commented on src/main/java/org/graylog/collector/file/GlobPathSet.java:
```diff
@@ -0,0 +1,167 @@
+        @Override
+        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) throws IOException {
+            // Skip /proc because it can throw some permission errors we cannot check for.
+            return dir.toString().startsWith("/proc") ? FileVisitResult.SKIP_SUBTREE : FileVisitResult.CONTINUE;
```
This means that for example `/procurement` will also be skipped? Maybe better check for equality with `/proc` *or* prefix `/proc/`.

bernd replied:
Good point!

joschi commented on src/main/java/org/graylog/collector/file/SinglePathSet.java:
```diff
@@ -0,0 +1,76 @@
+
+    @Override
+    public Set<Path> getPaths() throws IOException {
+        return Files.exists(path) ? ImmutableSet.of(path) : ImmutableSet.<Path>of();
```
Minor nit-pick: Use `Collections#emptySet()` to avoid unnecessary object allocation.

bernd replied:
Aye, thanks!

joschi commented:
Splitting up into `path-glob-root` and `path-glob-pattern` *feels* a little unnatural to me. Isn't there a way to use a config setting instead that covers all cases?

bernd committed:
Improve /proc check in GlobbingFileVisitor.

Use equals() instead of startsWith() to avoid skipping all directories
starting with /proc.

bernd committed:
Use Collections.emptySet() instead of ImmutableSet.of().

Avoids object allocation.

bernd commented:
> Splitting up into `path-glob-root` and `path-glob-pattern` feels a little unnatural to me. Isn't there a way to use a config setting instead that covers all cases?

I agree. It would be nicer to be able to use `path = "/var/log/**/*.log"` and be done with it. The problem is that we have to split it up to get a root path for the file observer and also that we have to escape parts of the path on Windows.

On Windows we have to escape the directory separator `\` because it is a special character for globbing. (`C:\test\logs\*.log` needs to become `C:\\test\\logs\\*.log`) So now we have to decide what part of the path string needs to be escaped and which not.

bernd commented:
Updated after comments.

bernd committed:
Wording for log message when files do not exist yet.

joschi commented:
LGTM. :+1:
